Reload config update pinned_images

roman-kiselenko opened the pull request:
#### What type of PR is this?

/kind bug
/kind feature

#### What this PR does / why we need it:

Updates pinned images list on config reload

#### Which issue(s) this PR fixes:

Fixes #7954 

#### Special notes for your reviewer:



#### Does this PR introduce a user-facing change?

None

```release-note

Updates pinned images list on config reload

```

/cc @haircommander @sohankunkerkar @hexfusion @kwilczynski

sohankunkerkar commented on server/server.go:
```diff
@@ -583,6 +583,19 @@ func (s *Server) startReloadWatcher(ctx context.Context) {
+	if err := s.config.Reload(); err != nil {
+		return err
+	}
+	if len(s.config.PinnedImages) > 0 {
```
I think instead of checking the length, how about we change the logic of `ReloadPinnedImages` to return a boolean type? This will ensure that `UpdateImageCache` is called only when `ReloadPinnedImages` returns true indicating an update to the list of pinned_images.

sohankunkerkar reviewed (commented):
@roman-kiselenko Thanks for working on this issue. There's already an integration test for this to cover the use case of updating the pinned_images when cri-o restarts, could you add another one when cri-o reloads?

kwilczynski commented:
/ok-to-test

kwilczynski commented:
/assign roman-kiselenko

sohankunkerkar commented:
How about we update the image cache in `reloadConfigByHUP` only when there's an update to the list of pinned_images, instead of passing the `updateImageCache func([]string)` to `Reload`?
```go
func (c *Config) ReloadPinnedImages(newConfig *Config) bool {
	updatedPinnedImages := make([]string, len(newConfig.PinnedImages))
	updatedPinnedImageList := false

	for i, image := range newConfig.PinnedImages {
		if i < len(c.PinnedImages) && image == c.PinnedImages[i] {
			continue
		}
		updatedPinnedImages[i] = image
		updatedPinnedImageList = true
	}

	if updatedPinnedImageList {
		logrus.Infof("Updated new pinned images: %+v", updatedPinnedImages)
		c.PinnedImages = updatedPinnedImages
	}
	
	return updatedPinnedImageList
}
```

And then in reloadConfigByHUP, use this:
```go
// reloadConfigByHUP reload the config and UpdateImageCache
func (s *Server) reloadConfigByHUP() error {
	if err := s.config.Reload(); err != nil {
		return err
	}
	if if s.config.ReloadPinnedImages() {
		if err := s.StorageImageServer().UpdateImageCache(s.config.PinnedImages); err != nil {
			return err
		}
	}
	return nil
}
```

roman-kiselenko commented:
> How about we update the image cache in `reloadConfigByHUP` only when there's an update to the list of pinned_images, instead of passing the `updateImageCache func([]string)` to `Reload`?
> 
> ```go
> func (c *Config) ReloadPinnedImages(newConfig *Config) bool {
> 	updatedPinnedImages := make([]string, len(newConfig.PinnedImages))
> 	updatedPinnedImageList := false
> 
> 	for i, image := range newConfig.PinnedImages {
> 		if i < len(c.PinnedImages) && image == c.PinnedImages[i] {
> 			continue
> 		}
> 		updatedPinnedImages[i] = image
> 		updatedPinnedImageList = true
> 	}
> 
> 	if updatedPinnedImageList {
> 		logrus.Infof("Updated new pinned images: %+v", updatedPinnedImages)
> 		c.PinnedImages = updatedPinnedImages
> 	}
> 	
> 	return updatedPinnedImageList
> }
> ```
> 
> And then in reloadConfigByHUP, use this:
> 
> ```go
> // reloadConfigByHUP reload the config and UpdateImageCache
> func (s *Server) reloadConfigByHUP() error {
> 	if err := s.config.Reload(); err != nil {
> 		return err
> 	}
> 	if if s.config.ReloadPinnedImages() {
> 		if err := s.StorageImageServer().UpdateImageCache(s.config.PinnedImages); err != nil {
> 			return err
> 		}
> 	}
> 	return nil
> }
> ```

cool, I'm going to use your solutions, thanks 🙏

sohankunkerkar commented:
> The only thing here is how to pass `newConfig` variable. That the reason for passing `updateImageCache func([]string)` to `Reload()` 🤔
> 
> ```go
> s.config.ReloadPinnedImages() 
> ```


hmm... I think you need to wrap `ReloadPinnedImages` in another function, maybe?

```go
func (c *Config) IsPinnedImagesUpdated() bool {
	return c.ReloadPinnedImages(c)
}
```

haircommander commented on pkg/config/reload.go:
```diff
@@ -13,8 +13,8 @@ import (
-// configuration directory.
-func (c *Config) Reload() error {
+// configuration directory. Method takes a function to update PinnedImages as parameter.
+func (c *Config) Reload(updateImageCache func([]string) error) error {
```
is it possible to use the `_inject_test.go` pattern where we can override for the tests specifically by calling a hidden setter?

roman-kiselenko replied:
I've changed a bit of implementation and used another way of passing `Config.` PTAL

roman-kiselenko commented:
> > The only thing here is how to pass `newConfig` variable. That the reason for passing `updateImageCache func([]string)` to `Reload()` 🤔
> > ```go
> > s.config.ReloadPinnedImages() 
> > ```
> 
> hmm... I think you need to wrap `ReloadPinnedImages` in another function, maybe?
> 
> ```go
> func (c *Config) IsPinnedImagesUpdated() bool {
> 	return c.ReloadPinnedImages(c)
> }
> ```

I've changed a bit of implementation and used another way of passing Config. PTAL

roman-kiselenko commented:
> > The only thing here is how to pass `newConfig` variable. That the reason for passing `updateImageCache func([]string)` to `Reload()` 🤔
> > ```go
> > s.config.ReloadPinnedImages() 
> > ```
> 
> hmm... I think you need to wrap `ReloadPinnedImages` in another function, maybe?
> 
> ```go
> func (c *Config) IsPinnedImagesUpdated() bool {
> 	return c.ReloadPinnedImages(c)
> }
> ```

The `Reload()` function contains a `newConfig` variable with fresh PinnedImages. I will rewrite the `Reload()` logic a little so the `PinnedImages` attribute will update appropriately.

roman-kiselenko commented:
/retest

sohankunkerkar commented:
/retest

sohankunkerkar reviewed (approved):
Changes LGTM
/lgtm

sohankunkerkar commented:
@cri-o/cri-o-maintainers PTAL

haircommander commented on internal/storage/image.go:
```diff
@@ -897,6 +901,26 @@ func (st nativeStorageTransport) ResolveReference(ref types.ImageReference) (typ
 	return istorage.ResolveReference(ref)
 }
 
+// UpdateImageCache invalidate imageCache if the pinned image from config exists;
```
nit: invalidates*

haircommander commented on test/reload_config.bats:
```diff
@@ -230,3 +230,13 @@ EOF
+	# when
+	reload_crio
+	#then
+	wait_for_log 'Updated new pinned images'
```
I think this test could do a `crictl images -o json` and find the pinned images, then do so after we do the reload and make sure the images are updated

roman-kiselenko replied:
This comment led me to understand that the current implementation is wrong, and it doesn't update the pinned attribute :) 

Thanks 🙏

haircommander replied:
I trust the config hadn't pinned anything else, but just to future proof, can you also do the opposite check before we reload_crio?

haircommander commented on server/server.go:
```diff
@@ -573,10 +573,17 @@ func (s *Server) startReloadWatcher(ctx context.Context) {
 				logrus.Errorf("Unable to reload configuration: %v", err)
 				continue
 			}
+			if s.config.IsPinnedImagesUpdated(newConfig) {
```
It's not clear to me how this condition will be true if we're replacing the config in place in Reload(). won't newConfig == s.config at this point?

roman-kiselenko replied:
The `Reload()` does not replace the config; it only updates the `newConfig` variable, and all the changes vanish with returns.

https://github.com/cri-o/cri-o/blob/0ae4367ef3b337894b13cd199b007653a5c2b6be/pkg/config/reload.go#L26-L38

haircommander replied:
wait but we do replace the c.config in reload: https://github.com/cri-o/cri-o/blob/0ae4367ef3b337894b13cd199b007653a5c2b6be/pkg/config/reload.go#L148-L156 we absorb the newConfig, and then take select changes into the c.config (old config) so how would the newConfig and c.config ever differ  WRT pinned images?

roman-kiselenko replied:
We're merging the old config pinned image list with the new pinned image list, and if there are changes, we update the pinned image list in `imageService`. Because we're relying on the returned state from `ReloadPinnedImages` we can perform `imageService` update only outside the Reload method.

hexfusion commented:
Just checking in on this how are we feeling about the current direction @haircommander @sohankunkerkar

haircommander commented on internal/storage/image.go:
```diff
@@ -897,6 +900,11 @@ func (st nativeStorageTransport) ResolveReference(ref types.ImageReference) (typ
 
+// UpdatePinnedImagesList updates pinned images list in imageService.
+func (svc *imageService) UpdatePinnedImagesList(configPinnedImage []string) {
+	svc.regexForPinnedImages = append(svc.regexForPinnedImages, CompileRegexpsForPinnedImages(configPinnedImage)...)
```
I'm not sure we should append here. Wouldnt' we just want to overwrite the slice? someone could theoretically unpin an image and we'd want it to qualify for GC then

hexfusion commented on pkg/config/reload.go:
```diff
@@ -13,20 +13,20 @@ import (
-// configuration directory.
-func (c *Config) Reload() error {
+// configuration directory. Returns updated Config struct.
+func (c *Config) Reload() (*Config, error) {
```
this feels a little awkward since Reload is a method of Config.

hexfusion commented on pkg/config/reload_test.go:
```diff
@@ -35,9 +35,10 @@ var _ = t.Describe("Config", func() {
 
 			// When
-			err := sut.Reload()
+			cfg, err := sut.Reload()
```
In this use case sut is aleady reference to `Config` right?

roman-kiselenko commented:
@haircommander @sohankunkerkar @hexfusion 

The trick with `Reload()` and return config confuses readers. I've changed the implementation to make it more readable. 

The `Reload()` method now updates the config struct, and fields `PinnedImages` and `PauseImage` are passed to `UpdatePinnedImagesList`.
https://github.com/cri-o/cri-o/blob/fbb899bb89a73917da0649a5c702964637454b32/internal/storage/image.go#L857-L859 

Both fields are updated in ImageService; hence, we must update both when HUP happens.

Also, I've added a test to catch some corner cases.

@cri-o/cri-o-maintainers PTAL

haircommander commented:
Yes I like this better! Good rewrite, and thanks for your patience
/approve
Tagging another @cri-o/cri-o-maintainers for the final review

haircommander commented on server/server.go:
```diff
@@ -577,6 +577,12 @@ func (s *Server) startReloadWatcher(ctx context.Context) {
+			// ImageServer compiles the list with regex for both
+			// pinned and sandbox/pause images, we need to update them
+			s.StorageImageServer().UpdatePinnedImagesList(
+				s.config.PauseImage,
```
Totally non blocking nit: I think i'd prefer combining the config fields here instead of passing two args

haircommander commented on internal/storage/image.go:
```diff
@@ -897,6 +900,14 @@ func (st nativeStorageTransport) ResolveReference(ref types.ImageReference) (typ
+// UpdatePinnedImagesList updates pinned images list in imageService.
+func (svc *imageService) UpdatePinnedImagesList(pauseImage string, pinnedImages []string) {
+	if pauseImage != "" {
+		pinnedImages = append(pinnedImages, pauseImage)
```
If pause image is changed to ", what happens here? I forget, do we fallback to anything in sandbox_run_linux.go, and if so should we include it here?

roman-kiselenko replied:
That's a good catch; I'll update it. :pray:

roman-kiselenko replied:
I'm a bit confused; could you elaborate? 

AFAIU, the pause image in runPodSandbox comes from the config. Hence, it will be updated on HUP.  

https://github.com/cri-o/cri-o/blob/fbb899bb89a73917da0649a5c702964637454b32/server/sandbox_run_linux.go#L461-L464

roman-kiselenko replied:
> If pause image is changed to "

This will not happen because the empty value in the config will not be parsed

> Unable to reload configuration: invalid reference format

Hence, `DefaultPauseImage` will always be in the list of `PinnedImages` and as `PauseImage`.

sohankunkerkar replied:
Could you add an integration test for this?

sohankunkerkar replied:
Just to make sure everything works as expected.

roman-kiselenko replied:
I've implemented that checking here https://github.com/cri-o/cri-o/pull/7976/files#diff-4ea762d405a8752ac8b728bfbe94b8ee5f962013a1e65d4e5d20208556418447R255-R257

haircommander reviewed (approved):
Actually a couple of notes but non blocking

roman-kiselenko commented:
/retest

sohankunkerkar commented on pkg/config/reload.go:
```diff
@@ -73,6 +75,7 @@ func (c *Config) Reload() error {
 	}
 	cdi.GetRegistry(cdi.WithSpecDirs(newConfig.CDISpecDirs...))
 
+	cdi.GetRegistry(cdi.WithSpecDirs(newConfig.CDISpecDirs...))
```
why do we have two entries here?

roman-kiselenko replied:
ah, bad rebase thx

hexfusion commented on pkg/config/reload.go:
```diff
@@ -143,17 +146,23 @@ func (c *Config) ReloadPauseImage(newConfig *Config) error {
 }
 
 // ReloadPinnedImages updates the PinnedImages with the provided `newConfig`.
```
optional: Might make sense to add a few words on what the bool this public method now returns indicates?

roman-kiselenko replied:
Thx, I've rewritten this method and removed the `bool` return. Not log print happens in the same way as in other methods with usage of `logConfig()`

sohankunkerkar commented on server/server.go:
```diff
@@ -577,6 +577,9 @@ func (s *Server) startReloadWatcher(ctx context.Context) {
 			}
+			// ImageServer compiles the list with regex for both
+			// pinned and sandbox/pause images, we need to update them
+			s.StorageImageServer().UpdatePinnedImagesList(append(s.config.PinnedImages, s.config.PauseImage))
```
With this change, you can drop this code: https://github.com/cri-o/cri-o/blob/main/internal/storage/image.go#L857

sohankunkerkar replied:
ah, right!

hexfusion commented on pkg/config/reload.go:
```diff
@@ -143,17 +144,23 @@ func (c *Config) ReloadPauseImage(newConfig *Config) error {
 }
 
 // ReloadPinnedImages updates the PinnedImages with the provided `newConfig`.
+// The method print log in case of any updates.
```
:+1:

roman-kiselenko commented:
The code has been changed a lot since last approve, I've re-request review 🙏

hexfusion reviewed (commented):
I am not a maintainer of this code but this approach looks much better :+1:

haircommander commented:
/approve
Still, @sohankunkerkar for the final lgtm?

hexfusion commented:
The integrations might be flakes but I think you need to fix the shfmt linter
```
make shfmt
```
probably looking for whitespace

sohankunkerkar commented:
> The integrations might be flakes but I think you need to fix the shfmt linter
> 
> ```
> make shfmt
> ```
> 
> probably looking for whitespace

@roman-kiselenko Thanks for your patience. One last fix before `/lgtm`

hexfusion commented:
fedora-kata test is failing due to rate limiting
> rateLimitExceeded

roman-kiselenko commented:
> fedora-kata test is failing due to rate limiting
> 
> > rateLimitExceeded

[integration / userns / amd64](https://github.com/cri-o/cri-o/actions/runs/8737513570/job/23974871352?pr=7976#step:13:416) - failed and I cant find a reason. 🤔 

[integration / conmon / arm64](https://github.com/cri-o/cri-o/actions/runs/8737513570/job/23974870432?pr=7976#step:13:352) - NRI tests are flaky

sohankunkerkar reviewed (approved):
/lgtm

hexfusion commented:
/test ci-fedora-kata

sohankunkerkar commented:
>last failure might not of been a flake

Yup, it's a legit one.

hexfusion commented:
cc @roman-kiselenko PTAL

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/7976/pull-ci-cri-o-cri-o-main-ci-fedora-kata/1780966933546405888/artifacts/fedora-kata/cri-o-gather/artifacts/testout.txt
```
not ok 216 reload config should update 'pinned_images' and 'pause_image' is pinned automatically
# (in test file reload_config.bats, line 260)
#   `[[ "$output" == *"wasm"* ]]' failed
```

roman-kiselenko commented:
> cc @roman-kiselenko PTAL
> 
> https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/7976/pull-ci-cri-o-cri-o-main-ci-fedora-kata/1780966933546405888/artifacts/fedora-kata/cri-o-gather/artifacts/testout.txt
> ```
> not ok 216 reload config should update 'pinned_images' and 'pause_image' is pinned automatically
> # (in test file reload_config.bats, line 260)
> #   `[[ "$output" == *"wasm"* ]]' failed
> ```

I can't reproduce it locally :sad:

going to poke around logs for a while

roman-kiselenko commented:
> I think this failure is only for Kata; the other two integration tests are looking fine

Do you think we can disable them for Kata?

I can't find the reason why it failed.

sohankunkerkar commented:
@littlejawa  Do you have any idea here?

sohankunkerkar commented:
```sh
 level=error msg="Getting /opt/kata/bin/containerd-shim-kata-v2 OCI runtime features failed: io.containerd.kata.v2: shim namespace cannot be empty: exit status 1" file="config/config.go:1306"
```

roman-kiselenko commented:
> ```shell
>  level=error msg="Getting /opt/kata/bin/containerd-shim-kata-v2 OCI runtime features failed: io.containerd.kata.v2: shim namespace cannot be empty: exit status 1" file="config/config.go:1306"
> ```

That looks like a problem; I've found this https://github.com/kata-containers/kata-containers/issues/541 

TBH I don't know how CRI-O works with Kata.  😞

roman-kiselenko commented:
Also [image pull & list disable on Kata](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/8019/pull-ci-cri-o-cri-o-main-ci-fedora-kata/1780041107673976832/artifacts/fedora-kata/cri-o-gather/artifacts/testout.txt#:~:text=ok%2030%20image%20pull%20and%20list%20by%20manifest,and%20manifest%20list%20digest%27%20not%20working%20in%20kata)

sohankunkerkar commented on contrib/test/ci/vars.yml:
```diff
@@ -44,7 +44,7 @@ ssh_user: "{{ lookup('env','USER') }}"
 
 # variables for testing with the kata-containers runtime
-kata_version: "3.1.3"
+kata_version: "3.3.0"
```
There's already a [PR](https://github.com/cri-o/cri-o/pull/7958) to test the new version of Kata.

roman-kiselenko replied:
oh, thanks, I didn't spot this.

sohankunkerkar commented:
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/7976/pull-ci-cri-o-cri-o-main-ci-fedora-integration/1782419051419537408/artifacts/fedora-integration/cri-o-gather/artifacts/testout.txt

I think [this](https://github.com/cri-o/cri-o/pull/8057) should solve the issue.

roman-kiselenko commented:
/retest

hexfusion commented:
Have we tested inside a kata instance to verify what is going on? Does the test provide the details of how to do this?

hexfusion commented:
cc @roman-kiselenko

roman-kiselenko commented:
> We can skip this test here: https://github.com/cri-o/cri-o/blob/main/contrib/test/ci/vars.yml#L101 and open an issue to track the test failure.

I've skipped this test for `kata.` I'm going to create an issue for this after a successful merge.

roman-kiselenko commented:
/retest

roman-kiselenko commented:
/retest

sohankunkerkar reviewed (approved):
/lgtm

sohankunkerkar commented:
/retest

hexfusion commented:
/cherry-pick release-1.29

hexfusion commented:
cc @sohankunkerkar can we get a cherry pick to 1.29?

hexfusion commented:
thanks for seeing this through @roman-kiselenko

sohankunkerkar commented:
/cherry-pick release-1.29

littlejawa commented:
Hey there,
Sorry for being late on this one, I was offline for a while. I'll have a look at the issue and try to understand what's wrong on the kata side.

roman-kiselenko commented:
> Hey there,
> Sorry for being late on this one; I was offline for a while. I'll have a look at the issue and try to understand what's wrong on the kata side.

hey, thank you for your time.

We have an umbrella issue here https://github.com/cri-o/cri-o/issues/8074
